# 모든 assay에 대한 smiles찾기

In [3]:
!pip install pubchempy

  Using cached PubChemPy-1.0.4.tar.gz (29 kB)
  Created wheel for pubchempy: filename=PubChemPy-1.0.4-py3-none-any.whl size=13834 sha256=d2d29ded42239e6939da7367de4571c821520f4178e380b6307d6bf5f020b367
  Stored in directory: /home/kch/.cache/pip/wheels/b0/8c/ba/3b00b89931153bf5a4eaa8e73bd1b0319a879cc45175326854
Successfully built pubchempy


In [19]:
import pandas as pd
_335_ToxcastAssays_IntendedTarget_zcoreFiltered = pd.read_csv('./Data/335_ToxcastAssays_IntendedTarget_zcoreFiltered.csv',index_col = "Name")

In [20]:
assays_data_335 = _335_ToxcastAssays_IntendedTarget_zcoreFiltered

In [21]:
assays_data_335

TOX21_AhR_LUC_Agonist TOX21_Aromatase_Inhibition  \
Name                                                            
100-00-5                         N                          N   
1000-05-1                        N                          N   
10001-13-5                       N                          N   
100-01-6                         N                          N   
100-02-7                         N                          N   
...                            ...                        ...   
NOCAS_903892                   NaN                        NaN   
NOCAS_903893                   NaN                        NaN   
NOCAS_903894                   NaN                        NaN   
NOCAS_903895                   NaN                        NaN   
NOCAS_903896                   NaN                        NaN   

             TOX21_VDR_BLA_agonist_ratio TOX21_VDR_BLA_antagonist_ratio  \
Name                                                                      
100-00-5                               N                              N   
1000-05-1                              N                              N   
10001-13-5                             N                              N   
100-01-6                               N                              N   
100-02-7                               N                              N   
...                                  ...                            ...   
NOCAS_903892                         NaN                            NaN   
NOCAS_903893                         NaN                            NaN   
NOCAS_903894                         NaN                            NaN   
NOCAS_903895                         NaN                            NaN   
NOCAS_903896                         NaN                            NaN   

             ATG_Ahr_CIS_up NVS_ADME_hCYP2C19 NVS_ADME_hCYP19A1  \
Name                                                              
100-00-5                  N               NaN               NaN   
1000-05-1                 N               NaN               NaN   
10001-13-5              NaN               NaN               NaN   
100-01-6                  N       6.449239794               NaN   
100-02-7                  N                 N               NaN   
...                     ...               ...               ...   
NOCAS_903892              N               NaN               NaN   
NOCAS_903893              N               NaN               NaN   
NOCAS_903894              N               NaN               NaN   
NOCAS_903895              N               NaN               NaN   
NOCAS_903896              N               NaN               NaN   

             NVS_ADME_hCYP2C9 NVS_ADME_hCYP1A1 NVS_ADME_hCYP1A2  ...  \
Name                                                             ...   
100-00-5                  NaN              NaN              NaN  ...   
1000-05-1                 NaN              NaN              NaN  ...   
10001-13-5                NaN              NaN              NaN  ...   
100-01-6                  NaN              NaN              NaN  ...   
100-02-7                  NaN              NaN              NaN  ...   
...                       ...              ...              ...  ...   
NOCAS_903892              NaN              NaN              NaN  ...   
NOCAS_903893              NaN              NaN              NaN  ...   
NOCAS_903894              NaN              NaN              NaN  ...   
NOCAS_903895              NaN              NaN              NaN  ...   
NOCAS_903896              NaN              NaN              NaN  ...   

             TOX21_PPARg_BLA_Agonist_ratio TOX21_PPARd_BLA_agonist_ratio  \
Name                                                                       
100-00-5                                 N                             N   
1000-05-1                                N                             N   
10001-13-5                               N                             N   
10

In [22]:
df = assays_data_335
count_nan = 0
smiles_list = []
CASN_nosmiles = []
CASN_nosmiles_index = []
for i, casn in enumerate(df.index) : #while all cas numbers
  try :
    smiles_list.append(CASN_smiles_dict[casn])
  except : 
    smiles_list.append(float('nan'))
    count_nan+=1
    CASN_nosmiles.append(casn)
    CASN_nosmiles_index.append(i)
# print(count_nan)
# print(CASN_nosmiles)
# print(CASN_nosmiles_index)
for casn in CASN_nosmiles : 
  print(casn)

100-00-5
1000-05-1
10001-13-5
100-01-6
100-02-7
10004-44-1
100-06-1
100-10-7
100-11-8
100-14-1
100-15-2
10016-20-3
100-18-5
100-19-6
100-20-9
100-21-0
100-22-1
10022-28-3
10022-31-8
10023-54-8
10024-93-8
100-25-4
10025-74-8
10025-82-8
10025-87-3
10025-91-9
10026-13-8
10026-24-1
1002-69-3
100-27-6
1002-84-2
100286-90-6
100-29-8
100299-08-9
10030-73-6
10032-15-2
10034-93-2
100-36-7
100-37-8
100-39-0
10039-54-0
100-40-3
10040-45-6
100-41-4
100-42-5
10042-59-8
10043-35-3
10043-52-4
100-43-6
100-44-7
100-46-9
100-47-0
100-48-1
10049-83-9
100-50-5
100-51-6
10052-47-8
100-52-7
100-53-8
100-54-9
100-55-0
100-56-1
10061-01-5
100-61-8
100-63-0
100-64-1
100643-71-8
100-65-2
100678-32-8
100-69-6
100-70-9
1007-28-9
100-74-3
100-75-4
100-76-5
100784-20-1
100-80-1
10081-67-1
100-85-6
1008-65-7
100-86-7
1008-72-6
100-88-9
10094-34-5
1009-61-6
100-97-0
1009-93-4
101001-34-7
10101-97-0
101-02-0
10102-18-8
10102-45-1
101-05-3
10108-64-2
101-10-0
101-11-1
101-14-4
101152-94-7
101-18-8
10118-90-8
101200-48

In [28]:
count_nan

9224

In [25]:
print(len(smiles_list))
print(len(CASN_nosmiles))

9224


EPA comptox dashboard에서 batch단위로 smiles 찾아서 채우기 위해 5000개씩 나눔

In [41]:
len(CASN_nosmiles)

9224

In [42]:
textfile = open("./Data/CASN_batch.txt", "w")
for element in CASN_nosmiles:
    textfile.write(element + "\n")
textfile.close()

The SMILES downloaded with this selection are for the chemical structure associated with the substance displayed in a particular record. They can be considered to be "substance-level SMILES". Other types of SMILES are available including "QSAR-ready" and "MS-Ready" forms. QSAR-Ready SMILES are those that were used in order to perform the OPERA predictions (processed to desalt, remove stereo and split multi-component chemicals into individual components). MS-Ready SMILES has similar processing but uses a different list of salt components to retain mappings.

MS-Ready SMILES are the SMILES representations of the desalted, de-isotoped, stereo-neutral forms of chemical structures associated with particular chemical substances. One MS-Ready SMILES can map to multiple individual chemical substances.

QSAR-Ready SMILES are the SMILES representations of the desalted, de-isotoped, stereo-neutral forms of chemical structures associated with particular chemical substances. One QSAR-Ready SMILES can map to multiple individual chemical substances. They differ from MS-Ready SMILES based on a more restricted set of salt counterions primarily.

In [37]:
# !gsutil cp 335_all_assay.csv gs://chem_dsrc/P0_Toxcast/Data/
!gsutil cp gs://chem_dsrc/P0_Toxcast/Data/335_all_assay.csv ./

Copying gs://chem_dsrc/P0_Toxcast/Data/335_all_assay.csv...
- [1 files][835.5 KiB/835.5 KiB]                                                
Operation completed over 1 objects/835.5 KiB.                                    


In [112]:
batchSearched = pd.read_csv("./Data/comtoxBatchSearched.csv",index_col = "INPUT")

In [113]:
batchSearched

FOUND_BY          DTXSID  \
INPUT                                   
100-00-5        CASRN   DTXSID5020281   
1000-05-1       CASRN   DTXSID8044759   
10001-13-5      CASRN   DTXSID4046288   
100-01-6        CASRN   DTXSID8020961   
100-02-7        CASRN   DTXSID0021834   
...               ...             ...   
NOCAS_903892    CASRN  DTXSID70903892   
NOCAS_903893    CASRN  DTXSID30903893   
NOCAS_903894    CASRN  DTXSID90903894   
NOCAS_903895    CASRN  DTXSID50903895   
NOCAS_903896    CASRN  DTXSID10903896   

                                       PREFERRED_NAME  \
INPUT                                                   
100-00-5                      1-Chloro-4-nitrobenzene   
1000-05-1     1,1,3,3,5,5,7,7-Octamethyltetrasiloxane   
10001-13-5                                   Pexantel   
100-01-6                           4-Nitrobenzenamine   
100-02-7                                4-Nitrophenol   
...                                               ...   
NOCAS_903892                     EPAMED_Water_Set5-33   
NOCAS_903893                     EPAMED_Water_Set5-34   
NOCAS_903894                     EPAMED_Water_Set5-35   
NOCAS_903895                     EPAMED_Water_Set5-36   
NOCAS_903896                     EPAMED_Water_Set5-37   

                                                 SMILES  \
INPUT                                                     
100-00-5                    [O-][N+](=O)C1=CC=C(Cl)C=C1   
1000-05-1     C[SiH](C)O[Si](C)(C)O[Si](C)(C)O[SiH](C)C   
10001-13-5                     CN1CCN(CC1)C(=O)C1CCCCC1   
100-01-6                     NC1=CC=C(C=C1)[N+]([O-])=O   
100-02-7                     OC1=CC=C(C=C1)[N+]([O-])=O   
...                                                 ...   
NOCAS_903892                                        NaN   
NOCAS_903893                                        NaN   
NOCAS_903894                                        NaN   
NOCAS_903895                                        NaN   
NOCAS_903896                                        NaN   

                          MS_READY_SMILES  \
INPUT                                       
100-00-5      [O-][N+](=O)C1=CC=C(Cl)C=C1   
1000-05-1                             NaN   
10001-13-5       CN1CCN(CC1)C(=O)C1CCCCC1   
100-01-6       NC1=CC=C(C=C1)[N+]([O-])=O   
100-02-7       OC1=CC=C(C=C1)[N+]([O-])=O   
...                                   ...   
NOCAS_903892                          NaN   
NOCAS_903893                          NaN   
NOCAS_903894                          NaN   
NOCAS_903895                          NaN   
NOCAS_903896                          NaN   

                                      QSAR_READY_SMILES  
INPUT                                                    
100-00-5                    [O-][N+](=O)C1=CC=C(Cl)C=C1  
1000-05-1     C[SiH](C)O[Si](C)(C)O[Si](C)(C)O[SiH](C)C  
10001-13-5                     CN1CCN(CC1)C(=O)C1CCCCC1  
100-01-6                     NC1=CC=C(C=C1)[N+]([O-])=O  
100-02-7                     OC1=CC=C(C=C1)[N+]([O-])=O  
...                                                 ...  
NOCAS_903892                                        NaN  
NOCAS_903893                                        NaN  
NOCAS_903894                                        NaN  
NOCAS_903895                                        NaN  
NOCAS_903896                                        NaN  

[9257 rows x 6 columns]

In [ ]:
9224개의 casn을 넣었는데 9257개의 결과가 나왔다 
왜? --> 33개의 casn에 대해서 2개씩나옴. 

In [77]:
#Synonym mapped to two or more chemicals
casns_duplicated = batchSearched.index.value_counts()[:9257-9224].keys()

In [78]:
casn_duplicated_rows = batchSearched.loc[casns_duplicated]

In [63]:
casn_duplicated_rows.to_csv("./Data/casn_duplicated_rows.csv")

In [74]:
chnms_df = pd.read_csv("./Data/Chemical_Summary_190708.csv",index_col = "casn")

In [82]:
chnms_df

chid                  chnm          code dsstox_substance_id  \
casn                                                                           
60-35-5        20005             Acetamide        C60355       DTXSID7020005   
103-90-2       20006         Acetaminophen       C103902       DTXSID2020006   
968-81-0       20007         Acetohexamide       C968810       DTXSID7020007   
1975-05-08     20009          Acetonitrile        C75058       DTXSID7020009   
520-45-6       20014    Dehydroacetic acid       C520456       DTXSID6020014   
...              ...                   ...           ...                 ...   
NOCAS_903892  903892  EPAMED_Water_Set5-33  CNOCAS903892      DTXSID70903892   
NOCAS_903893  903893  EPAMED_Water_Set5-34  CNOCAS903893      DTXSID30903893   
NOCAS_903894  903894  EPAMED_Water_Set5-35  CNOCAS903894      DTXSID90903894   
NOCAS_903895  903895  EPAMED_Water_Set5-36  CNOCAS903895      DTXSID50903895   
NOCAS_903896  903896  EPAMED_Water_Set5-37  CNOCAS903896      DTXSID10903896   

                                                           clib  
casn                                                             
60-35-5       TOX21SL|TOXCAST|TOXCAST_EPATox21|TOXCAST_ph2|T...  
103-90-2      TOX21SL|TOXCAST|TOXCAST_EPATox21|TOXCAST_ph2|T...  
968-81-0      TOX21SL|TOXCAST|TOXCAST_EPATox21|TOXCAST_ph3|T...  
1975-05-08                                      TOX21SL|TOXCAST  
520-45-6      TOX21SL|TOXCAST|TOXCAST_e1k|TOXCAST_EPATox21|T...  
...                                                         ...  
NOCAS_903892                                                NaN  
NOCAS_903893                                                NaN  
NOCAS_903894                                                NaN  
NOCAS_903895                                                NaN  
NOCAS_903896                                                NaN  

[9224 rows x 5 columns]

In [104]:
9257-9224

33

In [85]:
dsstox_substance_ids =chnms_df.loc[casns_duplicated].dsstox_substance_id

In [ ]:
duplicated_casns_dsstox_substance_ids = casn_duplicated_rows.DTXSID

In [ ]:
dsstox_substance_ids

In [101]:
len(duplicated_casns_dsstox_substance_ids)

66

In [102]:
len(dsstox_substance_ids)

33

In [92]:
lst1 = list( duplicated_casns_dsstox_substance_ids )
lst2 = list( dsstox_substance_ids ) 
complement = list(set(lst1) - set(lst2))

In [103]:
len(complement)

33

In [114]:
batchSearched['INPUT'] = batchSearched.index
batchSearched.index = batchSearched.DTXSID

In [115]:
batchSearched

FOUND_BY          DTXSID  \
DTXSID                                    
DTXSID5020281     CASRN   DTXSID5020281   
DTXSID8044759     CASRN   DTXSID8044759   
DTXSID4046288     CASRN   DTXSID4046288   
DTXSID8020961     CASRN   DTXSID8020961   
DTXSID0021834     CASRN   DTXSID0021834   
...                 ...             ...   
DTXSID70903892    CASRN  DTXSID70903892   
DTXSID30903893    CASRN  DTXSID30903893   
DTXSID90903894    CASRN  DTXSID90903894   
DTXSID50903895    CASRN  DTXSID50903895   
DTXSID10903896    CASRN  DTXSID10903896   

                                         PREFERRED_NAME  \
DTXSID                                                    
DTXSID5020281                   1-Chloro-4-nitrobenzene   
DTXSID8044759   1,1,3,3,5,5,7,7-Octamethyltetrasiloxane   
DTXSID4046288                                  Pexantel   
DTXSID8020961                        4-Nitrobenzenamine   
DTXSID0021834                             4-Nitrophenol   
...                                                 ...   
DTXSID70903892                     EPAMED_Water_Set5-33   
DTXSID30903893                     EPAMED_Water_Set5-34   
DTXSID90903894                     EPAMED_Water_Set5-35   
DTXSID50903895                     EPAMED_Water_Set5-36   
DTXSID10903896                     EPAMED_Water_Set5-37   

                                                   SMILES  \
DTXSID                                                      
DTXSID5020281                 [O-][N+](=O)C1=CC=C(Cl)C=C1   
DTXSID8044759   C[SiH](C)O[Si](C)(C)O[Si](C)(C)O[SiH](C)C   
DTXSID4046288                    CN1CCN(CC1)C(=O)C1CCCCC1   
DTXSID8020961                  NC1=CC=C(C=C1)[N+]([O-])=O   
DTXSID0021834                  OC1=CC=C(C=C1)[N+]([O-])=O   
...                                                   ...   
DTXSID70903892                                        NaN   
DTXSID30903893                                        NaN   
DTXSID90903894                                        NaN   
DTXSID50903895                                        NaN   
DTXSID10903896                                        NaN   

                            MS_READY_SMILES  \
DTXSID                                        
DTXSID5020281   [O-][N+](=O)C1=CC=C(Cl)C=C1   
DTXSID8044759                           NaN   
DTXSID4046288      CN1CCN(CC1)C(=O)C1CCCCC1   
DTXSID8020961    NC1=CC=C(C=C1)[N+]([O-])=O   
DTXSID0021834    OC1=CC=C(C=C1)[N+]([O-])=O   
...                                     ...   
DTXSID70903892                          NaN   
DTXSID30903893                          NaN   
DTXSID90903894                          NaN   
DTXSID50903895                          NaN   
DTXSID10903896                          NaN   

                                        QSAR_READY_SMILES         INPUT  
DTXSID                                                                   
DTXSID5020281                 [O-][N+](=O)C1=CC=C(Cl)C=C1      100-00-5  
DTXSID8044759   C[SiH](C)O[Si](C)(C)O[Si](C)(C)O[SiH](C)C     1000-05-1  
DTXSID4046288                    CN1CCN(CC1)C(=O)C1CCCCC1    10001-13-5  
DTXSID8020961                  NC1=CC=C(C=C1)[N+]([O-])=O      100-01-6  
DTXSID0021834                  OC1=CC=C(C=C1)[N+]([O-])=O      100-02-7  
...                                                   ...           ...  
DTXSID70903892                                        NaN  NOCAS_903892  
DTXSID30903893                                        NaN  NOCAS_903893  
DTXSID90903894                                        NaN  NOCAS_903894  
DTXSID50903895                                        NaN  NOCAS_903895  
DTXSID10903896                                        NaN  NOCAS_903896  

[9257 rows x 7 columns]

In [128]:
batchSearched.INPUT.value_counts()[:33]

85976-54-1     2
1996-12-08     2
4292-19-7      2
57-30-7        2
25389-94-0     2
52549-17-4     2
36362-97-7     2
2795-39-3      2
26027-38-3     2
142-83-6       2
111-02-4       2
90357-06-5     2
113-42-8       2
79-34-5        2
87-69-4        2
111974-72-2    2
14897-39-3     2
122-40-7       2
95-76-1        2
137-30-4       2
695-53-4       2
51630-58-1     2
82385-42-0     2
360-70-3       2
87463-91-0     2
89-25-8        2
1989-02-01     2
127-68-4       2
104-40-5       2
6319-84-2      2
208993-54-8    2
120-29-6       2
3810-74-0      2
Name: INPUT, dtype: int64

In [129]:
duplicated_DTXSIDs = batchSearched.index.value_counts()[:8].keys()

In [ ]:
# CASN가 달라도 같은 물질인 경우가 있음(이 경우 DTXSID0020074 이 같음)

In [131]:
batchSearched.loc[duplicated_DTXSIDs]

FOUND_BY  \
DTXSID                                                             
DTXSID0020074  Integrated Source CAS-RN - WARNING: Synonym ma...   
DTXSID0020074                                              CASRN   
DTXSID9023546                                              CASRN   
DTXSID9023546  Integrated Source CAS-RN - WARNING: Synonym ma...   
DTXSID3031864                                              CASRN   
DTXSID3031864  Integrated Source CAS-RN - WARNING: Synonym ma...   
DTXSID7020479                                              CASRN   
DTXSID7020479  Integrated Source CAS-RN - WARNING: Synonym ma...   
DTXSID5021251                                              CASRN   
DTXSID5021251  Integrated Source CAS-RN - WARNING: Synonym ma...   
DTXSID5046986                                              CASRN   
DTXSID5046986  Integrated Source CAS-RN - WARNING: Synonym ma...   
DTXSID5029055  Integrated Source CAS-RN - WARNING: Synonym ma...   
DTXSID5029055                                              CASRN   
DTXSID5021122                                              CASRN   
DTXSID5021122  Integrated Source CAS-RN - WARNING: Synonym ma...   

                      DTXSID                PREFERRED_NAME  \
DTXSID                                                       
DTXSID0020074  DTXSID0020074                    Gabapentin   
DTXSID0020074  DTXSID0020074                    Gabapentin   
DTXSID9023546  DTXSID9023546                    Quetiapine   
DTXSID9023546  DTXSID9023546                    Quetiapine   
DTXSID3031864  DTXSID3031864  Perfluorooctanesulfonic acid   
DTXSID3031864  DTXSID3031864  Perfluorooctanesulfonic acid   
DTXSID7020479  DTXSID7020479                    Dimethoate   
DTXSID7020479  DTXSID7020479                    Dimethoate   
DTXSID5021251  DTXSID5021251                     Saccharin   
DTXSID5021251  DTXSID5021251                     Saccharin   
DTXSID5046986  DTXSID5046986              dl-Tartaric acid   
DTXSID5046986  DTXSID5046986              dl-Tartaric acid   
DTXSID5029055  DTXSID5029055       4-Nonylphenol, branched   
DTXSID5029055  DTXSID5029055       4-Nonylphenol, branched   
DTXSID5021122  DTXSID5021122                 Phenobarbital   
DTXSID5021122  DTXSID5021122                 Phenobarbital   

                                                          SMILES  \
DTXSID                                                             
DTXSID0020074                                NCC1(CC(O)=O)CCCCC1   
DTXSID0020074                                NCC1(CC(O)=O)CCCCC1   
DTXSID9023546     OCCOCCN1CCN(CC1)C1=NC2=C(SC3=C1C=CC=C3)C=CC=C2   
DTXSID9023546     OCCOCCN1CCN(CC1)C1=NC2=C(SC3=C1C=CC=C3)C=CC=C2   
DTXSID3031864  OS(=O)(=O)C(F)(F)C(F)(F)C(F)(F)C(F)(F)C(F)(F)C...   
DTXSID3031864  OS(=O)(=O)C(F)(F)C(F)(F)C(F)(F)C(F)(F)C(F)(F)C...   
DTXSID7020479                               CNC(=O)CSP(=S)(OC)OC   
DTXSID7020479                               CNC(=O)CSP(=S)(OC)OC   
DTXSID5021251                         O=C1NS(=O)(=O)C2=C1C=CC=C2   
DTXSID5021251                         O=C1NS(=O)(=O)C2=C1C=CC=C2   
DTXSID5046986                      O[C@@H]([C@H](O)C(O)=O)C(O)=O   
DTXSID5046986                      O[C@@H]([C@H](O)C(O)=O)C(O)=O   
DTXSID5029055                                                NaN   
DTXSID5029055                                                NaN   
DTXSID5021122                  CCC1(C(=O)NC(=O)NC1=O)C1=CC=CC=C1   
DTXSID5021122                  CCC1(C(=O)NC(=O)NC1=O)C1=CC=CC=C1   

                                                 MS_READY_SMILES  \
DTXSID                                                             
DTXSID0020074                                NCC1(CC(O)=O)CCCCC1   
DTXSID0020074                                NCC1(CC(O)=O)CCCCC1   
DTXSID9023546     OCCOCCN1CCN(CC1)C1=NC2=C(SC3=C1C=CC=C3)C=CC=C2   
DTXSID9023546     OCCOCCN1CCN(CC1)C1=NC2=C(SC3=C1C=CC=C3)C=CC=C2   
DTXSID3031864  OS(=O)(=O)C(F)(F)C(F)(F)C(F)(F)C(F)(F)C(F)(F)C...   
DTXSID3031864

In [ ]:
batchSearched

In [121]:
batchSearched.loc[complement].index.value_counts()

DTXSID0020074      2
DTXSID3031864      2
DTXSID5021251      2
DTXSID7020479      2
DTXSID5046986      2
DTXSID5029055      2
DTXSID5021122      2
DTXSID9023546      2
DTXSID101017940    1
DTXSID80859230     1
DTXSID00859151     1
DTXSID00904978     1
DTXSID70859224     1
DTXSID901017944    1
DTXSID2043797      1
DTXSID401017163    1
DTXSID6091550      1
DTXSID70861120     1
DTXSID901020244    1
DTXSID3058618      1
DTXSID3048079      1
DTXSID1032014      1
DTXSID4064767      1
DTXSID4023597      1
DTXSID501018140    1
DTXSID20859279     1
DTXSID00858720     1
DTXSID2041171      1
DTXSID101017364    1
DTXSID4051453      1
DTXSID30872595     1
DTXSID10859960     1
DTXSID9062142      1
Name: DTXSID, dtype: int64

In [109]:
batchSearched = batchSearched.drop(complement)

In [110]:
batchSearched

FOUND_BY          DTXSID  \
DTXSID                                    
DTXSID5020281     CASRN   DTXSID5020281   
DTXSID8044759     CASRN   DTXSID8044759   
DTXSID4046288     CASRN   DTXSID4046288   
DTXSID8020961     CASRN   DTXSID8020961   
DTXSID0021834     CASRN   DTXSID0021834   
...                 ...             ...   
DTXSID70903892    CASRN  DTXSID70903892   
DTXSID30903893    CASRN  DTXSID30903893   
DTXSID90903894    CASRN  DTXSID90903894   
DTXSID50903895    CASRN  DTXSID50903895   
DTXSID10903896    CASRN  DTXSID10903896   

                                         PREFERRED_NAME  \
DTXSID                                                    
DTXSID5020281                   1-Chloro-4-nitrobenzene   
DTXSID8044759   1,1,3,3,5,5,7,7-Octamethyltetrasiloxane   
DTXSID4046288                                  Pexantel   
DTXSID8020961                        4-Nitrobenzenamine   
DTXSID0021834                             4-Nitrophenol   
...                                                 ...   
DTXSID70903892                     EPAMED_Water_Set5-33   
DTXSID30903893                     EPAMED_Water_Set5-34   
DTXSID90903894                     EPAMED_Water_Set5-35   
DTXSID50903895                     EPAMED_Water_Set5-36   
DTXSID10903896                     EPAMED_Water_Set5-37   

                                                   SMILES  \
DTXSID                                                      
DTXSID5020281                 [O-][N+](=O)C1=CC=C(Cl)C=C1   
DTXSID8044759   C[SiH](C)O[Si](C)(C)O[Si](C)(C)O[SiH](C)C   
DTXSID4046288                    CN1CCN(CC1)C(=O)C1CCCCC1   
DTXSID8020961                  NC1=CC=C(C=C1)[N+]([O-])=O   
DTXSID0021834                  OC1=CC=C(C=C1)[N+]([O-])=O   
...                                                   ...   
DTXSID70903892                                        NaN   
DTXSID30903893                                        NaN   
DTXSID90903894                                        NaN   
DTXSID50903895                                        NaN   
DTXSID10903896                                        NaN   

                            MS_READY_SMILES  \
DTXSID                                        
DTXSID5020281   [O-][N+](=O)C1=CC=C(Cl)C=C1   
DTXSID8044759                           NaN   
DTXSID4046288      CN1CCN(CC1)C(=O)C1CCCCC1   
DTXSID8020961    NC1=CC=C(C=C1)[N+]([O-])=O   
DTXSID0021834    OC1=CC=C(C=C1)[N+]([O-])=O   
...                                     ...   
DTXSID70903892                          NaN   
DTXSID30903893                          NaN   
DTXSID90903894                          NaN   
DTXSID50903895                          NaN   
DTXSID10903896                          NaN   

                                        QSAR_READY_SMILES         INPUT  
DTXSID                                                                   
DTXSID5020281                 [O-][N+](=O)C1=CC=C(Cl)C=C1      100-00-5  
DTXSID8044759   C[SiH](C)O[Si](C)(C)O[Si](C)(C)O[SiH](C)C     1000-05-1  
DTXSID4046288                    CN1CCN(CC1)C(=O)C1CCCCC1    10001-13-5  
DTXSID8020961                  NC1=CC=C(C=C1)[N+]([O-])=O      100-01-6  
DTXSID0021834                  OC1=CC=C(C=C1)[N+]([O-])=O      100-02-7  
...                                                   ...           ...  
DTXSID70903892                                        NaN  NOCAS_903892  
DTXSID30903893                                        NaN  NOCAS_903893  
DTXSID90903894                                        NaN  NOCAS_903894  
DTXSID50903895                                        NaN  NOCAS_903895  
DTXSID10903896                                        NaN  NOCAS_903896  

[9216 rows x 7 columns]

최종적으로 9216개가 중복되지 않음 --> 이렇게 되면 assay데이터에서도 지워야함.

In [141]:
batchSearched.SMILES.value_counts()[:12]

CCC1(C(=O)NC(=O)NC1=O)C1=CC=CC=C1                                                    2
OC[C@@H](O)[C@@H](O)[C@H](O)[C@@H](O)C=O                                             2
NCC1(CC(O)=O)CCCCC1                                                                  2
CC(=O)O[C@@H]1C[C@@H]2CC[C@@]1(C)C2(C)C                                              2
OCCOCCN1CCN(CC1)C1=NC2=C(SC3=C1C=CC=C3)C=CC=C2                                       2
CNC(=O)CSP(=S)(OC)OC                                                                 2
O[C@@H]([C@H](O)C(O)=O)C(O)=O                                                        2
O=C1NS(=O)(=O)C2=C1C=CC=C2                                                           2
CC(C)[C@@H]1CC[C@@H](C)C[C@H]1O                                                      2
[H][C@@]12CC[C@@](O)(C#C)[C@@]1(CC)CC[C@]1([H])[C@@]3([H])CCC(=O)C=C3CC[C@@]21[H]    2
OS(=O)(=O)C(F)(F)C(F)(F)C(F)(F)C(F)(F)C(F)(F)C(F)(F)C(F)(F)C(F)(F)F                  2
CC(C)CCC[C@@H](C)CCC[C@@H](C)CCC[C@]1(C)CCC

In [159]:
batchSearchedSMILES = pd.DataFrame(batchSearched.SMILES)

In [165]:
batchSearchedSMILES.index = batchSearchedSMILES['SMILES']

In [167]:
batchSearchedSMILES = batchSearchedSMILES.dropna()

In [168]:
batchSearchedSMILES

SMILES
SMILES                                                                                               
[O-][N+](=O)C1=CC=C(Cl)C=C1                                               [O-][N+](=O)C1=CC=C(Cl)C=C1
C[SiH](C)O[Si](C)(C)O[Si](C)(C)O[SiH](C)C                   C[SiH](C)O[Si](C)(C)O[Si](C)(C)O[SiH](C)C
CN1CCN(CC1)C(=O)C1CCCCC1                                                     CN1CCN(CC1)C(=O)C1CCCCC1
NC1=CC=C(C=C1)[N+]([O-])=O                                                 NC1=CC=C(C=C1)[N+]([O-])=O
OC1=CC=C(C=C1)[N+]([O-])=O                                                 OC1=CC=C(C=C1)[N+]([O-])=O
...                                                                                               ...
NC1=CC=C(C(O)=O)C(O)=C1.CN(C)CCOC1=CC=CC=C1CC1=...  NC1=CC=C(C(O)=O)C(O)=C1.CN(C)CCOC1=CC=CC=C1CC1...
CC(=O)NC1=NC(=O)N(C=C1)C1COC(COC(=O)C2=CC=CC=C2)O1  CC(=O)NC1=NC(=O)N(C=C1)C1COC(COC(=O)C2=CC=CC=C...
OC(=O)\C=C\C(O)=O.CCN(CC)CCC1=CNC2=CC=CC=C12.CC...  OC(=O)\C=C\C(O)=O.CCN(CC)CCC1=CNC2=CC=CC=C12.C...
COCC(C)N(C(=O)CS(O)(=O)=O)C1=CSC=C1C                             COCC(C)N(C(=O)CS(O)(=O)=O)C1=CSC=C1C
CN1CC2=C(N[C@H](CC(O)=O)C1=O)C=CC(=C2)C(=O)N1CC...  CN1CC2=C(N[C@H](CC(O)=O)C1=O)C=CC(=C2)C(=O)N1C...

[8702 rows x 1 columns]

In [142]:
assay_name = "335_all_assay"
path = "./"
searched_df = pd.read_csv(path + assay_name+".csv")

In [148]:
searched_df

INPUT FOUND_BY          DTXSID  \
0         100-00-5   CAS-RN   DTXSID5020281   
1        1000-05-1   CAS-RN   DTXSID8044759   
2       10001-13-5   CAS-RN   DTXSID4046288   
3         100-01-6   CAS-RN   DTXSID8020961   
4         100-02-7   CAS-RN   DTXSID0021834   
...            ...      ...             ...   
9219  NOCAS_903892   CAS-RN  DTXSID70903892   
9220  NOCAS_903893   CAS-RN  DTXSID30903893   
9221  NOCAS_903894   CAS-RN  DTXSID90903894   
9222  NOCAS_903895   CAS-RN  DTXSID50903895   
9223  NOCAS_903896   CAS-RN  DTXSID10903896   

                               PREFERRED_NAME  \
0                     1-Chloro-4-nitrobenzene   
1     1,1,3,3,5,5,7,7-Octamethyltetrasiloxane   
2                                    Pexantel   
3                              4-Nitroaniline   
4                               4-Nitrophenol   
...                                       ...   
9219                     EPAMED_Water_Set5-33   
9220                     EPAMED_Water_Set5-34   
9221                     EPAMED_Water_Set5-35   
9222                     EPAMED_Water_Set5-36   
9223                     EPAMED_Water_Set5-37   

                                         SMILES  
0                   [O-][N+](=O)C1=CC=C(Cl)C=C1  
1     C[SiH](C)O[Si](C)(C)O[Si](C)(C)O[SiH](C)C  
2                      CN1CCN(CC1)C(=O)C1CCCCC1  
3                    NC1=CC=C(C=C1)[N+]([O-])=O  
4                    OC1=CC=C(C=C1)[N+]([O-])=O  
...                                         ...  
9219                                          -  
9220                                          -  
9221                                          -  
9222                                          -  
9223                                          -  

[9224 rows x 5 columns]

In [154]:
searched_dfSMILES = pd.DataFrame(searched_df.SMILES)

In [156]:
searched_dfSMILES.index = searched_dfSMILES.SMILES

In [157]:
searched_dfSMILES.drop('-')

SMILES
SMILES                                                                                               
[O-][N+](=O)C1=CC=C(Cl)C=C1                                               [O-][N+](=O)C1=CC=C(Cl)C=C1
C[SiH](C)O[Si](C)(C)O[Si](C)(C)O[SiH](C)C                   C[SiH](C)O[Si](C)(C)O[Si](C)(C)O[SiH](C)C
CN1CCN(CC1)C(=O)C1CCCCC1                                                     CN1CCN(CC1)C(=O)C1CCCCC1
NC1=CC=C(C=C1)[N+]([O-])=O                                                 NC1=CC=C(C=C1)[N+]([O-])=O
OC1=CC=C(C=C1)[N+]([O-])=O                                                 OC1=CC=C(C=C1)[N+]([O-])=O
...                                                                                               ...
NC1=CC=C(C(O)=O)C(O)=C1.CN(C)CCOC1=CC=CC=C1CC1=...  NC1=CC=C(C(O)=O)C(O)=C1.CN(C)CCOC1=CC=CC=C1CC1...
CC(=O)NC1=NC(=O)N(C=C1)C1COC(COC(=O)C2=CC=CC=C2)O1  CC(=O)NC1=NC(=O)N(C=C1)C1COC(COC(=O)C2=CC=CC=C...
OC(=O)\C=C\C(O)=O.CCN(CC)CCC1=CNC2=CC=CC=C12.CC...  OC(=O)\C=C\C(O)=O.CCN(CC)CCC1=CNC2=CC=CC=C12.C...
COCC(C)N(C(=O)CS(O)(=O)=O)C1=CSC=C1C                             COCC(C)N(C(=O)CS(O)(=O)=O)C1=CSC=C1C
CN1CC2=C(N[C@H](CC(O)=O)C1=O)C=CC(=C2)C(=O)N1CC...  CN1CC2=C(N[C@H](CC(O)=O)C1=O)C=CC(=C2)C(=O)N1C...

[8659 rows x 1 columns]

In [ ]:
searched_df = pd.read_csv(path + assay_name+".csv") # engine="python", encoding = "utf-8" 하면 .iterrows()에서 i가 for문의 index가 아니라.. row가 되버림..


In [ ]:
#1-1) EPA comptox dashboard에서 batch단위로 smiles 찾아서 채우기
smiles_searched_1= {}#찾은 smiles 딕셔러니
remove_index = [] #CASN_nosmiles, CASN_nosmiles_index에서 찾은것들은 제외하기 위한 인덱스
for i, row in searched_df.iterrows() : 
  if row[-1] != '-' : 
    smiles_searched_1[i] = row['SMILES'] #qsar ready-smiles로 한다.
    remove_index.append(i)
count_nan = count_nan - len(smiles_searched_1)
print("count_nan : ",count_nan)
#찾은 smiles 채우기
for key in smiles_searched_1.keys() : 
  smiles_list[ CASN_nosmiles_index[key] ] = smiles_searched_1[key]
#CASN_nosmiles과 CASN_nosmiles_index에서 찾은 것들은 지우기
CASN_nosmiles_2 = []
CASN_nosmiles_index_2 = []
remove_index_list = smiles_searched_1.keys()
for i in range(0,len(CASN_nosmiles)) : 
  if i in remove_index_list : 
    pass
  else : 
    CASN_nosmiles_2.append(CASN_nosmiles[i])
    CASN_nosmiles_index_2.append(CASN_nosmiles_index[i])
print( pd.Series( smiles_list).isnull().sum() )
print(len(CASN_nosmiles_2))
print(len(CASN_nosmiles_index_2))


# 1-2) PubchemPy패키지 이용하여 채우기
import pubchempy as pcp

searched_df_2 = [] 

for casn in CASN_nosmiles_2 : 
  # cmpnd = pcp.get_compounds(casn, 'name') #PubChem ID cid가 출력된다. 
  try : 
    cmpnd = pcp.get_compounds(casn, 'name') #PubChem ID cid가 출력된다. \
    searched_df_2.append(cmpnd[0].canonical_smiles)
  except : 
    searched_df_2.append('-')

smiles_searched_2 ={}
remove_index = [] #CASN_nosmiles_2, CASN_nosmiles_index_2에서 찾은것들은 제외하기 위한 인덱스
for i, row in enumerate(searched_df_2) : 
  if row != '-' : 
    smiles_searched_2[i] = row
    remove_index.append(i)
count_nan = count_nan - len(smiles_searched_2)
print("count_nan : ", count_nan)
#찾은 smiles채우기
for key in smiles_searched_2.keys() : 
  smiles_list[ CASN_nosmiles_index_2[key] ] = smiles_searched_2[key]
#CASN_nosmiles_2과 CASN_nosmiles_index_2에서 찾은 것들은 지우기
CASN_nosmiles_3 = []
CASN_nosmiles_index_3 = []
remove_index_list = smiles_searched_2.keys()
for i in range(0,len(CASN_nosmiles_2)) : 
  if i in remove_index_list : 
    pass
  else : 
    CASN_nosmiles_3.append(CASN_nosmiles_2[i])
    CASN_nosmiles_index_3.append(CASN_nosmiles_index_2[i])
print(pd.Series(smiles_list).isnull().sum())
print(len(CASN_nosmiles_3))
print(len(CASN_nosmiles_index_3))

#1-3) CIRconvert메서드 이용하여 채우기
from urllib.request import urlopen

def CIRconvert(ids):
    try:
        url = 'http://cactus.nci.nih.gov/chemical/structure/' + ids + '/smiles'
        ans = urlopen(url).read().decode('utf8')
        return ans
    except:
        return 'Did not work'


searched_df_3 = [] 

for casn in CASN_nosmiles_3 : 
  smiles = CIRconvert(casn)
  if smiles != 'Did not work' : 
      searched_df_3.append(smiles)
  else : 
      searched_df_3.append('-')

smiles_searched_3 ={}
remove_index = [] #CASN_nosmiles_3, CASN_nosmiles_index_3에서 찾은것들은 제외하기 위한 인덱스
for i, row in enumerate(searched_df_3) : 
  if row != '-' : 
    smiles_searched_3[i] = row
    remove_index.append(i)

count_nan = count_nan - len(smiles_searched_3)
print("count_nan : ",count_nan)

for key in smiles_searched_3.keys() : 
  smiles_list[ CASN_nosmiles_index_3[key] ] = smiles_searched_3[key]

CASN_nosmiles_4 = []
CASN_nosmiles_index_4 = []
remove_index_list = smiles_searched_3.keys()
for i in range(0,len(CASN_nosmiles_3)) : 
  if i in remove_index_list : 
    pass
  else : 
    CASN_nosmiles_4.append(CASN_nosmiles_3[i])
    CASN_nosmiles_index_4.append(CASN_nosmiles_index_3[i])

print(pd.Series(smiles_list).isnull().sum())
print(len(CASN_nosmiles_4))
print(len(CASN_nosmiles_index_4))

print("smiles를 찾지 못한 cans 리스트 : " ,CASN_nosmiles_4)
# df = df.to_frame()
df.insert(1, 'smiles', smiles_list)
df.to_csv(path + assay_name+"_df.csv")

count_nan :  565
565
565
565
count_nan :  417
417
417
417


In [ ]:
df.insert(0, 'smiles', smiles_list)
df.to_csv(path + assay_name+"_df.csv")

In [ ]:
df

Name                                             smiles  ... NVS_NR_hPPARa
100-00-5                    [O-][N+](=O)C1=CC=C(Cl)C=C1  ...           NaN
1000-05-1     C[SiH](C)O[Si](C)(C)O[Si](C)(C)O[SiH](C)C  ...           NaN
10001-13-5                     CN1CCN(CC1)C(=O)C1CCCCC1  ...           NaN
100-01-6                     NC1=CC=C(C=C1)[N+]([O-])=O  ...           NaN
100-02-7                     OC1=CC=C(C=C1)[N+]([O-])=O  ...           NaN
...                                                 ...  ...           ...
NOCAS_903892                                        NaN  ...           NaN
NOCAS_903893                                        NaN  ...           NaN
NOCAS_903894                                        NaN  ...           NaN
NOCAS_903895                                        NaN  ...           NaN
NOCAS_903896                                        NaN  ...           NaN

[9224 rows x 336 columns]

In [ ]:
!gsutil cp *_df.csv gs://chem_dsrc/P0_Toxcast/Data/assay_df/

Copying file://335_all_assay_df.csv [Content-Type=text/csv]...
/ [1 files][  4.2 MiB/  4.2 MiB]                                                
Operation completed over 1 objects/4.2 MiB.                                      
